### PR TITLE
JetStream iterator

### PR DESCRIPTION
**Pull request by alyakimenko**
### Description

Implement JetStream iterator.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-s3/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: source/config.go [hunk 1]
```diff
@@ -71,20 +75,21 @@ func Parse(cfg map[string]string) (Config, error) {
...
-		bufferSize, err := strconv.Atoi(cfg[ConfigKeyBufferSize])
-		if err != nil {
-			return Config{}, fmt.Errorf("\"%s\" must be an integer", ConfigKeyBufferSize)
-		}
+	if err := parseBufferSize(cfg[ConfigKeyBufferSize], &sourceConfig); err != nil {
+		return Config{}, fmt.Errorf("parse buffer size: %w", err)
+	}
 
-		sourceConfig.BufferSize = bufferSize
+	if err := parseDeliverPolicy(cfg[ConfigKeyDeliverPolicy], &sourceConfig); err != nil {
```
Comment (maksenius) [inline source/config.go:87]: Like proposition maybe move those functions to config methods

File: source/jetstream/iterator.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+// HasNext checks is the iterator has messages.
+func (i *Iterator) HasNext(ctx context.Context) bool {
+	return len(i.messages) > 0
+}
+
+// Next returns the next record from the underlying messages channel.
+// It also appends messages to a unackMessages slice if the AckPolicy is not equal to AckNonePolicy.
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
```
Comment (maksenius) [inline source/jetstream/iterator.go:100]: Do we need add for loop?

File: source/jetstream/iterator.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	// remove acknowledged message from the slice
+	i.unackMessages = i.unackMessages[1:]
+
+	return nil
+}
+
+// Stop stops the Iterator, unsubscribes from a subject.
+func (i *Iterator) Stop() (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (maksenius) [inline source/jetstream/iterator.go:147]: Is it can be crush when we do teardown?

File: source/config.go (hunk 1 above)
Comment (alyakimenko) [reply source/config.go]: Done

File: source/jetstream/iterator.go (hunk 2 above)
Comment (alyakimenko) [reply source/jetstream/iterator.go:100]: I think we don't need it here. Cause we just need to read one record from the channel

File: source/jetstream/iterator.go (hunk 3 above)
Comment (alyakimenko) [reply source/jetstream/iterator.go]: In the case when the `messages` channel is already closed. But yeah, it's merely impossible here. I deleted the recover

Comment (alyakimenko) [reply source/jetstream/iterator.go]: Only if Conduit will call the `Stop` method twice, for some reason

File: source/source.go [hunk 4]
```diff
@@ -81,7 +82,22 @@ func (s *Source) Open(ctx context.Context, position sdk.Position) error {
...
+			SDKPosition:   position,
+			DeliverPolicy: s.config.DeliverPolicy,
+			AckPolicy:     s.config.AckPolicy,
+		})
+		if err != nil {
+			return fmt.Errorf("init jetstream iterator: %w", err)
+		}
+
+	default:
+		return ErrUnknownCommunicationMode
```
Comment (maksenius) [inline source/source.go:100]: please add to err s.config.Mode value

Comment (alyakimenko) [reply source/source.go]: Done

**Review by maksenius — APPROVED** [review]

**Pull request merged**